src/core/thread/mle_router.cpp

```diff
@@ -3309,7 +3309,9 @@ void MleRouter::RemoveRouterLink(Router &aRouter)
 
 void MleRouter::RemoveNeighbor(Neighbor &aNeighbor)
 {
-    if (&aNeighbor == &mParent)
+    VerifyOrExit(!aNeighbor.IsStateInvalid(), OT_NOOP);
```

Comment:
An invalid neighbor with RLOC16 == 0xFFFF will be incorrectly identified as a `Child` in `mChildTable` in the original code.

Reply:
Do we call `RemoveNeighbor()` on neighbors with state invalid?

Reply:
Yes, we do. Here is the place where we found that the `RemoveNeighbor` can be removing an invalid neighbor:
https://github.com/openthread/openthread/blob/e839100be10e7c3a4119936c2d0bdfaf11effa94/src/core/thread/mle_router.cpp#L2238-L2243

Reply:
Should we consider fully clearing the neighbor entry when setting to invalid so that we do not match on it?

Reply:
I was a little confused too. 
Per my understanding, an `Router` in `Invalid` state is not actually **INVALID**. An actual **INVALID** `Router` is the one with `RLOC16 == 0xFFFF`. The implementation of `RouterTable::GetNextEntry` has confirmed it too: 
```cpp
const Router *RouterTable::GetNextEntry(const Router *aRouter) const
{
    VerifyOrExit(aRouter != NULL, OT_NOOP);
    aRouter++;
    VerifyOrExit(aRouter < &mRouters[Mle::kMaxRouters], aRouter = NULL);
    VerifyOrExit(aRouter->GetRloc16() != 0xffff, aRouter = NULL);

exit:
    return aRouter;
}
```

So, a `Router` of `Invalid` state would just mean that the `Router` is totally unreachable, but all those link information is still sound. Maybe these information can be useful later?

Reply:
That's all correct.

Here is how I usually treat the types (sort of my cheat sheet :))
- A `Neighbor` is a node we can directly talk to. `InvalidState` means it is not a neighbor (i.e. can not directly talk to it)
- `RouterTable` gives list of all routers within Thread network. Some `Router`s may be our neighbor and some may not be (e.g, two hops away).  `Router` inherits from `Neighbor` and uses `State` to indicate this. 
- A `Child` (`ChidTable`) is the list of devices own children and a `Child` is always a `Neighbor` (invalid state mean it is no longer a child).

`RouterTable` provides `GetNeighbor()` vs `GetRouter()` methods, first one returns a `Neighbor` (state is not invalid) vs second one returning a `Router` entry (may not be a neighbor).  

`ChildTable` provides `FindChild()` methods, returned entry is always a neighbor (state is not invalid). Let's caller to define a `StateFilter` to find `Child` in specific states.

`Mle` provides methods to `GetNeighbor()` (matching rloc, ext address, etc), returned value should be always a neighbor (i.e. not in invalid state). May be an entry from `ChildTable` or  `RouterTable` or one of special entries `Parent` or `ParentCandidate`.

Reply:
Yes, that is a good summary @abtink .